lib.php

```diff
@@ -529,4 +529,62 @@ function (stdClass $warning) use ($userid) : bool {
     foreach ($updatedwarnings as $updatedwarning) {
         $DB->update_record('attendance_warning', $updatedwarning);
     }
+}
+
+/**
+ * This function receives a calendar event and returns the action associated with it.
+ *
+ * This is used by block_timeline in order to display the event appropriately. If null is returned then the event
+ * is not displayed on the block.
+ *
+ * @param calendar_event $event
+ * @param \core_calendar\action_factory $factory
+ * @param int $userid
+ * @return \core_calendar\local\event\entities\action_interface|\core_calendar\local\event\value_objects\action
+ */
+function mod_attendance_core_calendar_provide_event_action(calendar_event $event,
+                                                           \core_calendar\action_factory $factory,
+                                                           $userid = 0) {
+    global $DB;
+
+    $url = '/mod/attendance/attendance.php';
+    $param = [];
+    $actiontext = '';
+
+    $sql = 'SELECT cm.id cmid, s.sessdate, s.id, s.studentscanmark FROM {attendance_sessions} s '
+         . 'JOIN {attendance} a ON a.id = s.attendanceid '
+         . 'JOIN {course_modules} cm ON (cm.course = a.course AND cm.instance = a.id) '
+         . 'WHERE cm.module = (SELECT id FROM {modules} WHERE name = \'attendance\') '
+         . 'AND s.caleventid = ?';
+    $sess = $DB->get_record_sql($sql, [$event->id]);
+
+    $context = context_module::instance($sess->cmid);
+
+    if (has_capability('mod/attendance:takeattendances', $context) && $sess->sessdate < time()) {
+        $actiontext = get_string('takeattendance', 'mod_attendance');
+        $url = '/mod/attendance/take.php';
+        $param = ['id' => $sess->cmid, 'sessionid' => $sess->id, 'grouptype' => 0];
+    } else if (has_capability('mod/attendance:canbelisted', $context)) {
+        $sql = 'SELECT s.description FROM {attendance_log} l '
```

Comment:
this looks very wrong - it should be using the function attendance_can_student_mark()

Reply:
what i am trying to do here (line 567) is say "are you are a student". The timeline event / session should always show to the student but the action link will change depending.

If attendance has been taken (line 571) the status is shown. If the attendance has not been taken and the session is in the past then perhaps i should be using attendance_can_student_mark().

If they can mark their own attendance, the link to take.php should be shown (line 578).

Future sessions do not show the action link. 

Comments included below
`if (has_capability('mod/attendance:takeattendances', $context) && $sess->sessdate < time()) {
        // The user can take attendance and the session is in the past. Show link to take (take.php)     
        $actiontext = get_string('takeattendance', 'mod_attendance');
        $url = '/mod/attendance/take.php';
        $param = ['id' => $sess->cmid, 'sessionid' => $sess->id, 'grouptype' => 0];
    } else if (has_capability('mod/attendance:canbelisted', $context)) {
        // The user can have their attendance taken
        $sql = 'SELECT s.description FROM {attendance_log} l '
             . 'JOIN {attendance_statuses} s ON s.id = l.statusid '
             . 'WHERE sessionid = ? AND studentid = ?';
        $recorded = $DB->get_record_sql($sql, [$sess->id, $userid]);

        if ($recorded !== false) {
            //  Attendance has been taken. Show the status and link to the attendance report (view.php)
            $actiontext = $recorded->description;
            $url = '/mod/attendance/view.php';
            $param = ['id' => $sess->cmid];
        } else {
            // if student can mark own attendance and the session is in the past show link to take (attendance.php)
            $actiontext = $sess->studentscanmark && $sess->sessdate < time()
                        ? get_string('submitattendance', 'mod_attendance') : "";
            $param = ['sessid' => $sess->id];
        }
    }`

Reply:
yeah - the action link to take attendance for a student should only show when attendance_can_student_mark returns that it can be shown - take a look at the other examples where that function is used. You can still show the content of the event, but you should not be displaying a link to take.php unless attendance_can_student_mark says that the user can actually take attendance for that session.